File: src/backend/interface.ts
export type ParseSourceArg = {
  sourceCode: string;
};

export type RevealEntityArg = {
  entityId: string;
};

export type SourceRange = {
  start_line: number;
  start_column: number;
  end_line: number;
  end_column: number;
};

export type AstEntityFull = {
  properties: Array<[string, string]>;
  children: AstEntityLite[];
  source_range: SourceRange | null;
};

export type TauriCommands = {
  parse_source: {
    args: ParseSourceArg;
    return: null;
  };
  reveal_entity: {
    args: RevealEntityArg;
    return: AstEntityFull;
  };
};

export type AstEntityLite = {
  id: string;
  kind: string;
  display_name: string;
};

export type SourceCode = {
  path: string;
  content: string;
};

export type EventPayload = {
  "ast-ready": AstEntityLite;
  "file-picked": SourceCode;
};

File: src-tauri/src/interface.rs
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct AstEntityLite {
    pub id: String,
    pub kind: String,
    pub display_name: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct SourceRange {
    pub start_line: i64,
    pub start_column: i64,
    pub end_line: i64,
    pub end_column: i64,
}

#[derive(Debug, Clone, Serialize)]
pub struct AstEntityFull {
    pub properties: Vec<(String, String)>,
    pub children: Vec<AstEntityLite>,
    pub source_range: Option<SourceRange>,
}

#[derive(Debug, Clone, Serialize)]
pub struct SourceCode {
    pub path: String,
    pub content: String,
}
